sql: allow escape of any character for COPY

otan opened the pull request:
According to the PostgreSQL docs:
  Any other backslashed character that is not mentioned in the above
  table will be taken to represent itself.

Reflect that change by converting error conditions from decodeCopy into
interpreting the characters as is.

Resolves #52067.

Release note (bug fix): `COPY` previously did not allow a backslash of
any character other than the special table set. CockroachDB would error
in these cases. This is changed to allow in any character after a
backslash and interpret it to mean the character itself as per PostgreSQL
(e.g. now `\a` will be interpreted as `a`). Furthermore, non hex-digits
following a `\x` is now interpreted without the backslash, (e.g. `\xH` as
`xH`). Strings ending with a single backslash will use the backslash
(e.g. `x\\` is interpreted as `x\`).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52068)
<!-- Reviewable:end -->

otan committed:
sql: allow escape of any character for COPY

According to the PostgreSQL docs:
  Any other backslashed character that is not mentioned in the above
  table will be taken to represent itself.

Reflect that change by converting error conditions from decodeCopy into
interpreting the characters as is.

Release note (bug fix): `COPY` previously did not allow a backslash of
any character other than the special table set. CockroachDB would error
in these cases. This is changed to allow in any character after a
backslash and interpret it to mean the character itself as per PostgreSQL
(e.g. now `\a` will be interpreted as `a`). Furthermore, non hex-digits
following a `\x` is now interpreted without the backslash, (e.g. `\xH` as
`xH`). Strings ending with a single backslash will use the backslash
(e.g. `x\\` is interpreted as `x\`).

otan commented:
bors r=mjibson

otan commented:
bors r=mjibson
